src/server/src/main/java/io/cassandrareaper/storage/memory/ReplicaLockManagerWithTtl.java

```diff
@@ -0,0 +1,169 @@
+  private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+  private final long ttlSeconds; // 1 minute
+
+  public ReplicaLockManagerWithTtl(long ttlSeconds) {
+    this.ttlSeconds = ttlSeconds;
+    // Schedule cleanup of expired locks
+    scheduler.scheduleAtFixedRate(this::cleanupExpiredLocks, 1, 1, TimeUnit.SECONDS);
+  }
+
+  private String getReplicaLockKey(String replica, UUID runId) {
+    return replica + runId;
+  }
+
+  public boolean lockRunningRepairsForNodes(UUID runId, UUID segmentId, Set<String> replicas) {
+    lock.lock();
+    try {
+      long currentTime = System.currentTimeMillis();
+      // Check if any replica is already locked by another runId
+      for (String replica : replicas) {
+        LockInfo lockInfo = replicaLocks.get(getReplicaLockKey(replica, runId));
+        if (lockInfo != null && lockInfo.expirationTime > currentTime && lockInfo.runId.equals(runId)) {
+          return false; // Replica is locked by another runId and not expired
+        }
+      }
+
+      // Lock the replicas for the given runId and segmentId
+      long expirationTime = currentTime + (ttlSeconds * 1000);
+      for (String replica : replicas) {
+        replicaLocks.put(getReplicaLockKey(replica, runId), new LockInfo(runId, expirationTime));
+      }
+
+      // Update runId to segmentId mapping
+      runToSegmentLocks.computeIfAbsent(runId, k -> ConcurrentHashMap.newKeySet()).add(segmentId);
+      return true;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public boolean renewRunningRepairsForNodes(UUID runId, UUID segmentId, Set<String> replicas) {
+    lock.lock();
+    try {
+      long currentTime = System.currentTimeMillis();
+
+      // Check if all replicas are already locked by this runId
+      for (String replica : replicas) {
+        LockInfo lockInfo = replicaLocks.get(getReplicaLockKey(replica, runId));
+        if (lockInfo == null || !lockInfo.runId.equals(runId) || lockInfo.expirationTime <= currentTime) {
+          return false; // Some replica is not validly locked by this runId
+        }
+      }
+
+      // Renew the lock by extending the expiration time
+      long newExpirationTime = currentTime + (ttlSeconds * 1000);
+      for (String replica : replicas) {
+        replicaLocks.put(getReplicaLockKey(replica, runId), new LockInfo(runId, newExpirationTime));
+      }
+
+      // Ensure the segmentId is linked to the runId
+      runToSegmentLocks.computeIfAbsent(runId, k -> ConcurrentHashMap.newKeySet()).add(segmentId);
+      return true;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public boolean releaseRunningRepairsForNodes(UUID runId, UUID segmentId, Set<String> replicas) {
+    lock.lock();
+    try {
+      // Remove the lock for replicas
+      for (String replica : replicas) {
+        LockInfo lockInfo = replicaLocks.get(getReplicaLockKey(replica, runId));
+        if (lockInfo != null && lockInfo.runId.equals(runId)) {
```

Comment:
I disagree, since we're trying to lock multiple replicas at the same time, we need to protect the code from race conditions between checking the locks and taking the locks.
The alternative would be to check without a lock and then risk a failure in placing the replica lock (and deal with the exception).
Given that everything in the method happens in memory, it's going to be a quick lock.

Reply:
In the event of failure to acquire the lock, isn't it just a retry anyway? Again, I'm not committed to this, it is just a suggestion.

Reply:
Let me try to provide an example later on today. This requires some thinking through.